orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RecreateUpdateStrategy.groovy

```diff
@@ -0,0 +1,139 @@
+import com.netflix.spinnaker.orca.kato.pipeline.support.ResizeStrategy
+import com.netflix.spinnaker.orca.pipeline.model.Execution
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import com.netflix.spinnaker.orca.pipeline.model.SyntheticStageOwner
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.context.ApplicationContext
+import org.springframework.context.ApplicationContextAware
+import org.springframework.stereotype.Component
+
+import static com.netflix.spinnaker.orca.pipeline.StageDefinitionBuilder.newStage
+
+@Slf4j
+@Component
+class RecreateUpdateStrategy implements Strategy, ApplicationContextAware {
+  final String name = "recreateupdate"
+
+  @Autowired
+  ResizeServerGroupStage resizeServerGroupStage
+
+  @Autowired
+  DetermineTargetServerGroupStage determineTargetServerGroupStage
+
+  @Autowired
+  TargetServerGroupResolver targetServerGroupResolver
+
+  @Override
+  <T extends Execution<T>> List<Stage<T>> composeFlow(Stage<T> stage) {
+    def stages = []
+    def stageData = stage.mapTo(RecreateUpdateStrategyData)
+    def cleanupConfig = AbstractDeployStrategyStage.CleanupConfig.fromStage(stage)
+
+    Map baseContext = [
+      (cleanupConfig.location.singularType()): cleanupConfig.location.value,
+      cluster                                : cleanupConfig.cluster,
+      credentials                            : cleanupConfig.account,
+      cloudProvider                          : cleanupConfig.cloudProvider,
+    ]
+
+    stage.context.capacity = [
+      min    : 0,
+      max    : 0,
+      desired: 0
+    ]
+
+    def p = 0
+    def findContext = baseContext + [
+      target        : TargetServerGroup.Params.Target.current_asg_dynamic,
+      targetLocation: cleanupConfig.location,
+    ]
+
+    stages << newStage(
+      stage.execution,
+      determineTargetServerGroupStage.type,
+      "Determine Deployed Server Group",
+      findContext,
+      stage,
+      SyntheticStageOwner.STAGE_BEFORE
+    )
+
+    def source
+    try {
+      source = getSource(targetServerGroupResolver, stageData, baseContext)
+    } catch (Exception e) {
+      stages = []
+      return stages
+    }
+
+    def resizeContext = baseContext + [
+      target        : TargetServerGroup.Params.Target.current_asg_dynamic,
+      action        : ResizeStrategy.ResizeAction.scale_exact,
+      source        : source,
+      targetLocation: cleanupConfig.location,
+      kind          : stageData.kind,
+      capacity      : stage.context.capacity,
+      resizeType    : "exact",
+      scaleNum      : 0
+    ]
+
+    def resizeStage = newStage(
+      stage.execution,
+      resizeServerGroupStage.type,
+      "Scale down to $p% Desired Size(0)",
+      resizeContext,
+      stage,
+      SyntheticStageOwner.STAGE_BEFORE
+    )
+    stages << resizeStage
+
+    return stages
```

Comment:
I'm still not sure what this stage is doing or how it's supposed to work - all that's being inserted is a single resize to 0 stage. If this behaves like other deployment strategies it will deploy the new server group, then immediately scale it down to 0 instances. Is that the goal?

Reply:
In order to update the controller, it needs to terminate the existing pods.  Resize is to terminate running pods and allow k8s to recreate with new pods with new definition.

Reply:
Which kind is this for?

Reply:
https://github.com/chlung/orca/blob/2d0c83393e389a3a036706957232a1d37b115fd4/orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RecreateUpdateStrategy.groovy#L95

Reply:
Well that's any kind, which one in particular? StatefulSet, DaemonSet, Job, etc...

Reply:
StatefulSet and DaemonSet will need it to use this strategy.  This is only need it for k8s OnDelete stratey.  RollingUpdate strategy doesn't require it.

Reply:
Wait so this is only for OnDelete? RollingUpdate I don't think is supported for k8s at all

Reply:
that is correct.  RollingUpdate is all controlled by k8s.  It does not depend on OnDelete event to trigger pod creation.  Sorry, I need to update spec.
RollingUpdate:
DaemonSet : 1.6 and after (https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/)
StatefulSet: 1.7 and after (http://blog.kubernetes.io/2017/09/kubernetes-statefulsets-daemonsets.html)

Reply:
So if we don't need to wait for pod deletion, we don't need this strategy right?

Reply:
there is two use cases.
1. OnDelete - use delete pod to trigger create new pod. without deletion, the pod will still running old definition.
2. RollingUodate - no action required.

Is this what you asking?

Reply:
So when a user, without spinnaker, updates a statefulset/daemonset, they have to manually resize it down to 0 pods?

Reply:
If so - doesn't this stage need to resize the controller back to the original capacity? And why would someone want to use this if they can just rely on RollingUpdate? I think the OnDelete is meant for a different scenario...

Reply:
Resize is just an option.  The purpose is to delete these running pods in order to launch new pods.  Why user want to use OnDelete opetion?  It just another option that k8s provides, and we support it as well. Especially, those people running older version of k8s.  This is another option.  I believe that RollingUpdate  will be used in most people as long as they are running the up-to-date k8s version.

Reply:
We are starting with k8s 1.5 that onDelete is only available for us to use at that time.

Reply:
Yes.  The controller will resize back to whatever the new defined size that all pods will be launched with new definition.

Reply:
Ahh OK I understand, this makes sense now. I think you need a second resize stage after the first to ensure that you are back at full capacity after the strategy finishes, right?

Reply:
I did not use patch library instead that I replace the whole controller.spec.  So, it can cover the replicas and resize according to the new controller definition.  The pipeline should have all controller definition because this is a deploy stage.  If you see any issue not patch library, please share with me. thanks.

Reply:
What do you mean by "patch library"? I'm worried that this pipeline will always leave a server group at size 0 when it finishes.

Reply:
I don't really compare any changes and patch the controller.  I basically replace the controller.spec and make replace api call.  Ideally, the controller will automatically launch pods.  The only potential problem only if controller failed during update and leave the resize to 0.  If you feel that the extra resize will help user.  I can add it.  Please let me know.

Reply:
Well as-is, this stage will always leave it scaled down to 0, right? If instead that only happens when the user makes an error, that seems better, right?

Reply:
I can include extra resize after update finish.  I do want to walk through my flow one more time with you.
Define statefulset with replicas set to 4 for update.
1. Pre-update: set the replicas to 0.
2. Deploy the update.  The replicas will be 4.  It will launch 4 pods after successful deployment.
***3. Now, add another resize to ensure 2 failed and pod will remain the original size.  If 2 is completed successfully, k8s will ignore this resize because pod size is same as replicas defined.

If we all agree, I will implement accordingly.  thanks.

Reply:
If the only way these can be updated is by destroying, why can't we use the highlander strategy?

Reply:
If I am not wrong, the highlander will dlete controller and create a new copy.  This is not the purpose of OnDelete.  We only replace the original controller definition and launch new pods.

Reply:
These deployment strategies always start by creating a new server group though - have you disabled that here?

Reply:
Yes.  I don't use deployment.  I don't use these labels to control cluster disable as well.

Reply:
I'm not sure what you mean - can you show me a screenshot of how the clusters panel looks while this is running?